types/module/module.go

```diff
@@ -299,19 +305,40 @@ func (m *Manager) RegisterServices(cfg Configurator) {
 	}
 }
 
-// InitGenesis performs init genesis functionality for modules. Exactly one
-// module must return a non-empty validator set update to correctly initialize
-// the chain.
+func (m *Manager) UnmarshalGenesis(cdc codec.JSONCodec, genesisData map[string]json.RawMessage) map[string]proto.Message {
+	var wg sync.WaitGroup
+	genesisStates := map[string]proto.Message{}
+
+	for _, moduleName := range m.OrderInitGenesis {
+		if genesisData[moduleName] == nil {
+			continue
+		}
+		wg.Add(1)
+		go func(moduleName string) {
+			defer wg.Done()
+			m.mtx.Lock()
+			genesisStates[moduleName] = m.Modules[moduleName].UnmarshalGenesis(cdc, genesisData[moduleName])
+			m.mtx.Unlock()
```

Comment:
How are we getting any parallelism in this design? The entire unmarshal is under the same mutex?

```suggestion
			defer wg.Done()
			moduleGenesis := m.Modules[moduleName].UnmarshalGenesis(cdc, genesisData[moduleName])
			m.mtx.Lock()
			genesisStates[moduleName] = moduleGenesis
			m.mtx.Unlock()
```

Reply:
You're correct @ValarDragon -- great catch. The marshaling needs to be concurrent, which it wasn't before and thus wasn't buying us anything.